core/src/main/java/org/elasticsearch/search/aggregations/AggregatorFactories.java

```diff
@@ -251,4 +259,97 @@ private void resolvePipelineAggregatorOrder(Map<String, AggregatorFactory> aggFa
+                factoriesList.toArray(new AggregatorFactory[factoriesList.size()]), pipelineAggregatorFactoriesList);
+        boolean hasParent = in.readBoolean();
+        if (hasParent) {
+            AggregatorFactory parent = in.readAggregatorFactory();
+            aggregatorFactories.parent = parent;
+        }
+        return aggregatorFactories;
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeVInt(this.factories.length);
+        for (AggregatorFactory factory : factories) {
+            out.writeAggregatorFactory(factory);
+        }
+        out.writeVInt(this.pipelineAggregatorFactories.size());
+        for (PipelineAggregatorFactory factory : pipelineAggregatorFactories) {
+            out.writePipelineAggregatorFactory(factory);
+        }
+        boolean hasParent = parent != null;
+        out.writeBoolean(hasParent);
+        if (hasParent) {
+            out.writeAggregatorFactory(parent);
+        }
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
+        if (factories != null) {
+            for (AggregatorFactory subAgg : factories) {
+                subAgg.toXContent(builder, params);
+            }
+        }
+        if (pipelineAggregatorFactories != null) {
+            for (PipelineAggregatorFactory subAgg : pipelineAggregatorFactories) {
+                subAgg.toXContent(builder, params);
+            }
+        }
+        builder.endObject();
+        return builder;
+    }
+
+    @Override
+    public int hashCode() {
+        final int prime = 31;
+        int result = 1;
+        result = prime * result + Arrays.hashCode(factories);
+        result = prime * result + ((parent == null) ? 0 : parent.hashCode());
+        result = prime * result + ((pipelineAggregatorFactories == null) ? 0 : pipelineAggregatorFactories.hashCode());
+        return result;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj)
+            return true;
+        if (obj == null)
+            return false;
+        if (getClass() != obj.getClass())
+            if (!(obj.getClass().isAssignableFrom(getClass()) || getClass().isAssignableFrom(obj.getClass())))
+                return false;
+        AggregatorFactories other = (AggregatorFactories) obj;
+        if (!Arrays.equals(factories, other.factories))
+            return false;
+        if (parent == null) {
+            if (other.parent != null)
+                return false;
+        } else if (!parent.equals(other.parent))
+            return false;
+        if (pipelineAggregatorFactories == null) {
+            if (other.pipelineAggregatorFactories != null)
+                return false;
+        } else if (!pipelineAggregatorFactories.equals(other.pipelineAggregatorFactories))
+            return false;
+        return true;
```

Comment:
we need to make these equals impls simpler. Maybe we can remove the "this == obj" optimization, use Objects.equal to avoid null checks, and remove the isAssignableFrom for classes that are never supposed to be equal to sub-classes (all?)?